File: Flyweight.cpp
/*****
 * Flyweight
 * Enables a large number of fine-grained objects to be shared.
*****/

#include <iostream>
#include <memory>
#include <unordered_map>
#include <string>

using namespace std;


class Flyweight { public: virtual void method(const string& name) = 0; };

class ConcreteFlyweight : public Flyweight {
    string key;

    public:
        explicit ConcreteFlyweight(string state) : key(move(state)) {}
        void method(const string& name) override { cout << key << "-" << name << endl; }
};

class Factory {
    unordered_map<string, unique_ptr<Flyweight>> flyweights;

    public:
        Flyweight* getFlyweight(const string& key) {
            auto it = flyweights.find(key);
            if (it == flyweights.end()) {
                it = flyweights.emplace(key, make_unique<ConcreteFlyweight>(key)).first;
            }
            return it->second.get();
        }
};


int main() {
    Factory f;
    
    Flyweight* f1 = f.getFlyweight("A");
    Flyweight* f2 = f.getFlyweight("A");
    f1->method("1");
    f2->method("2");
    
    Flyweight* f3 = f.getFlyweight("B");
    f3->method("3");
    
    return 0;
}

File: Flyweight.java
/*****
 * Flyweight
 * Enables a large number of fine-grained objects to be shared.
*****/

import java.util.HashMap;
import java.util.Map;


interface Flyweight_ { void method(String name); }

class ConcreteFlyweight implements Flyweight_ {
    private String key;

    public ConcreteFlyweight(String state) { this.key = state; }

    @Override
    public void method(String name) { System.out.println(key + "-" + name); }
}

class Factory {
    private Map<String, Flyweight_> flyweights = new HashMap<>();

    public Flyweight_ getFlyweight(String key) {
        if (!flyweights.containsKey(key)) {
            flyweights.put(key, new ConcreteFlyweight(key));
        }
        return flyweights.get(key);
    }
}


public class Flyweight {
    public static void main(String[] args) {
        Factory f = new Factory();

        Flyweight_ f1 = f.getFlyweight("A");
        Flyweight_ f2 = f.getFlyweight("A");
        f1.method("1");
        f2.method("2");

        Flyweight_ f3 = f.getFlyweight("B");
        f3.method("3");
    }
}
